pkg/webhook/webhooks.go

```diff
@@ -0,0 +1,272 @@
+	if ar.Request.Operation != v1beta1.Delete && req.Kind.Kind != "PersistentVolumeClaim" {
+		response.Allowed = true
+		return response
+	}
+	glog.Infof("AdmissionReview for Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+
+	// TODO* use raw object once validation webhooks support DELETE request
+	// object as non nil value https://github.com/kubernetes/kubernetes/issues/66536
+	//var pvc corev1.PersistentVolumeClaim
+	//err := json.Unmarshal(req.Object.Raw, &pvc)
+	//if err != nil {
+	//	glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+	//	status.Allowed = false
+	//	status.Result = &metav1.Status{
+	//		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+	//		Message: err.Error(),
+	//	}
+	//	return status
+	//}
+
+	// fetch the pvc specifications
+	pvc, err := wh.KubeClient.CoreV1().PersistentVolumeClaims(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("error retrieving PVC: %v", err.Error()),
+		}
+		return response
+	}
+
+	if !validationRequired(ignoredNamespaces, &pvc.ObjectMeta) {
+		glog.Infof("Skipping validation for %s/%s due to policy check", pvc.Namespace, pvc.Name)
+		return &v1beta1.AdmissionResponse{
+			Allowed: true,
+		}
+	}
+
+	// construct source-volume label to list all the matched cstorVolumes
+	label := fmt.Sprintf("openebs.io/source-volume=%s", pvc.Spec.VolumeName)
+	listOptions := metav1.ListOptions{
+		LabelSelector: label,
+	}
+
+	// get the all CStorVolumes resources in openebs namespace or in pvc namespaces
+	// based on the source-volume label to verify if there is any clone volume exists.
+	// if source-volume label matches with name of PV, failed the pvc
+	// deletion operation.
+
+	var cStorVolumes *v1alpha1.CStorVolumeList
+	response.Allowed = true
+	cStorVolumes, err = wh.Clientset.OpenebsV1alpha1().CStorVolumes(OpenebsNamespace).List(listOptions)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			glog.V(4).Infof("Checking in cstorVolumes in PVC namespace %s", req.Namespace)
+			cStorVolumes, err = wh.Clientset.OpenebsV1alpha1().CStorVolumes(req.Namespace).List(listOptions)
+			if err != nil {
+				response.Allowed = false
+				response.Result = &metav1.Status{
+					Message: fmt.Sprintf("error retrieving CstorVolumes: %v", err.Error()),
+				}
+			}
+		}
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("error retrieving CstorVolumes: %v", err.Error()),
+		}
+		return response
+	}
+
+	if len(cStorVolumes.Items) != 0 {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: "PVC has cloned volume",
+			Message: fmt.Sprintf("pvc %q has one or more cloned volume(s)", pvc.Name),
```

Comment:
let us give the count of cstorvolumes here

Reply:
added the count at the end